Yank/mpi.py

```diff
@@ -0,0 +1,421 @@
+on_single_node
+    Decorator version of run_single_node.
+distribute
+    Map a task on a sequence of arguments on all the nodes.
+delay_termination
+    A context manager to delay the response to termination signals.
+delayed_termination
+    A decorator version of delay_termination.
+
+"""
+
+
+# ==============================================================================
+# GLOBAL IMPORTS
+# ==============================================================================
+
+import os
+import sys
+import signal
+import logging
+from contextlib import contextmanager
+
+# TODO drop this when we drop Python 2 support
+from openmoltools.utils import wraps_py2
+
+logger = logging.getLogger(__name__)
+
+# ==============================================================================
+# GLOBAL VARIABLES
+# ==============================================================================
+
+disable_mpi = False
+
+
+# ==============================================================================
+# MAIN FUNCTIONS
+# ==============================================================================
+
+def get_mpicomm():
+    """Retrieve the MPI communicator for this execution.
+
+    The function automatically detects if the program runs on MPI by checking
+    specific environment variables set by various MPI implementations. On
+    first execution, it modifies sys.excepthook and register a handler for
+    SIGINT, SIGTERM, SIGABRT to call Abort() to correctly terminate all
+    processes.
+
+    Returns
+    -------
+    mpicomm : mpi4py communicator or None
+        The communicator for this node, None if the program doesn't run
+        with MPI.
+
+    """
+    # If MPI execution is forcefully disabled, return None.
+    if disable_mpi:
+        return None
+
+    # If we have already initialized MPI, return the cached MPI communicator.
+    if get_mpicomm._is_initialized:
+        return get_mpicomm._mpicomm
+
+    # Check for environment variables set by mpirun. Variables are from
+    # http://docs.roguewave.com/threadspotter/2012.1/linux/manual_html/apas03.html
+    variables = ['PMI_RANK', 'OMPI_COMM_WORLD_RANK', 'OMPI_MCA_ns_nds_vpid',
+                 'PMI_ID', 'SLURM_PROCID', 'LAMRANK', 'MPI_RANKID',
+                 'MP_CHILD', 'MP_RANK', 'MPIRUN_RANK']
+    use_mpi = False
+    for var in variables:
+        if var in os.environ:
+            use_mpi = True
+            break
+
+    # Return None if we are not running on MPI.
+    if not use_mpi:
+        logger.debug('Cannot find MPI environment. MPI disabled.')
+        get_mpicomm._mpicomm = None
+        return get_mpicomm._mpicomm
+
+    # Initialize MPI
+    from mpi4py import MPI
+    MPI.COMM_WORLD.barrier()
+    mpicomm = MPI.COMM_WORLD
+
+    # Override sys.excepthook to abort MPI on exception
+    def mpi_excepthook(type, value, traceback):
+        sys.__excepthook__(type, value, traceback)
+        sys.stdout.flush()
+        sys.stderr.flush()
+        if mpicomm.size > 1:
```

Comment:
Good question! I may be wrong about this, and I honestly forgot the details, but when I implemented the `excepthook` I read somewhere that `MPI_Abort` shouldn't be used to gracefully exit the application, but it's there just to shut everything down when there's no hope for recovery. I added the `mpicomm.size > 1` check because if we have only 1 process, MPI should handle closure automatically, and we won't risk having MPI processes hanging around on the cluster.